### DISCUSSION TITLE
Parsing CRI-formatted logs from `file` source?

### DISCUSSION THREAD

**Discussion by baodrate** [vectordotdev/vector#16986] [Q&A]
Is it possible to use the [`kubernetes_log`](https://vector.dev/docs/reference/configuration/sources/kubernetes_logs/#configuration) parser with [`file`](https://vector.dev/docs/reference/configuration/sources/file/), without all the kubernetes runtime stuff? Like as a transform, perhaps?

Motivation: I'm using the [podman driver with nomad](https://developer.hashicorp.com/nomad/plugins/drivers/podman) which creates a log file in the CRI logging format. (i.e. `<timestamp> <stdout|stderr> <P|F> [data]`). This is already implemented within https://github.com/vectordotdev/vector/blob/master/src/sources/kubernetes_logs/parser/cri.rs, but is there a way to use this parser with the standard `file` source?

Workarounds considered:

* Using the [journald log driver for nomad-podman](https://developer.hashicorp.com/nomad/plugins/drivers/podman#driver-journald).
  * This is doable, but has additional overhead because logs are parsed and copied/written by journald, vector, and nomad. 
  * Not sure of the best way to annotate these journald logs with nomad metadata. For `file` logging see: https://github.com/mr-karan/nomad-vector-logger and https://github.com/input-output-hk/nomad-follower

* Manual parsing
  * Seems like it should be possible
    * the [multiline](https://vector.dev/docs/reference/configuration/sources/file/#multiline) option + regex is probably insufficient since the two streams and interrupt each other
    * seems like you should be able to carry enough state with [lua parsing](https://vector.dev/docs/reference/configuration/transforms/lua/) to do the job
  * I'd prefer to avoid re-implementing parsing since it's already handled very well natively by vector

**Comment by jszwedko** [comment]
Hey!

Yes, I think using the `multiline` options and regex parsing would be the most straightforward.

If you felt like contributing, I think CRI parsing could be moved to a [codec](https://github.com/vectordotdev/vector/tree/master/lib/codecs/src/decoding/format) and more easily used by the `file` and other sources. The `syslog` codec you see in there is a fairly similar example. There would also need to be a new [framing](https://github.com/vectordotdev/vector/tree/master/lib/codecs/src/decoding/framing) option that understands how to collapse multiline CRI logs or the existing `multiline` options would need to continue to be used even if a `codec` existed since codecs only see one "frame" at a time.

**Reply by baodrate** [reply]
Thanks for the guidance! I'll give the parsing a shot. Then I'll see about putting together a PR for moving the CRI codec.

For the multiline parsing, do you have any advice on separating the `stdout` and `stderr` streams? Lines from one stream can intersperse the lines from the other, so filtering would need to be done before aggregating multiple lines. Would the right approach be:

```
source:file -> transform:route -> [stdout] transform:reduce -> sink
                               -> [stderr] transform:reduce -> 
```

**Reply by jszwedko** [reply]
Aaah, that's a good note. Yeah, in that case, using `reduce` would be better so that you can split before merging lines.

**Comment by jmealo** [comment]
@baodrate How did you make out? I'm using the file source with kubernetes to avoid calling the Kubernetes API.
